File: experiments-local/experiment-15/runs/agent=omp_language=go_model=openrouter/tencent/hy3-preview_tooling=none/rep3/handlers.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"strconv"
	"strings"
	"time"
	"unicode/utf8"
)

// Handler holds the dependencies for HTTP handlers
type Handler struct {
	store *BookStore
}

// NewHandler creates a new Handler with the given BookStore
func NewHandler(store *BookStore) *Handler {
	return &Handler{store: store}
}

// RegisterRoutes registers all routes on the given ServeMux
func (h *Handler) RegisterRoutes(mux *http.ServeMux) {
	mux.HandleFunc("/health", h.handleHealth)
	mux.HandleFunc("/books", h.handleBooks)
	mux.HandleFunc("/books/", h.handleBookByID)
}

// handleHealth handles GET /health
func (h *Handler) handleHealth(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	resp := HealthResponse{
		Status: "healthy",
		Time:   time.Now().Format(time.RFC3339),
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(resp)
}

// handleBooks handles GET /books and POST /books
func (h *Handler) handleBooks(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case http.MethodGet:
		h.listBooks(w, r)
	case http.MethodPost:
		h.createBook(w, r)
	default:
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
	}
}

// handleBookByID handles GET /books/{id}, PUT /books/{id}, DELETE /books/{id}
func (h *Handler) handleBookByID(w http.ResponseWriter, r *http.Request) {
	// Extract ID from path: /books/{id}
	path := strings.TrimPrefix(r.URL.Path, "/books/")
	if path == "" {
		http.Error(w, "Book ID required", http.StatusBadRequest)
		return
	}

	id, err := strconv.ParseInt(path, 10, 64)
	if err != nil {
		h.writeError(w, "Invalid book ID", http.StatusBadRequest)
		return
	}

	switch r.Method {
	case http.MethodGet:
		h.getBook(w, r, id)
	case http.MethodPut:
		h.updateBook(w, r, id)
	case http.MethodDelete:
		h.deleteBook(w, r, id)
	default:
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
	}
}

// listBooks handles GET /books
func (h *Handler) listBooks(w http.ResponseWriter, r *http.Request) {
	authorFilter := r.URL.Query().Get("author")

	books, err := h.store.ListBooks(authorFilter)
	if err != nil {
		log.Printf("Error listing books: %v", err)
		h.writeError(w, "Internal server error", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(books)
}

// createBook handles POST /books
func (h *Handler) createBook(w http.ResponseWriter, r *http.Request) {
	var req CreateBookRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		h.writeError(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	// Validate input
	if err := validateCreateBookRequest(req); err != nil {
		h.writeError(w, err.Error(), http.StatusBadRequest)
		return
	}

	book, err := h.store.CreateBook(req)
	if err != nil {
		log.Printf("Error creating book: %v", err)
		h.writeError(w, "Internal server error", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(book)
}

// getBook handles GET /books/{id}
func (h *Handler) getBook(w http.ResponseWriter, r *http.Request, id int64) {
	book, err := h.store.GetBookByID(id)
	if err != nil {
		if strings.Contains(err.Error(), "not found") {
			h.writeError(w, err.Error(), http.StatusNotFound)
		} else {
			log.Printf("Error getting book: %v", err)
			h.writeError(w, "Internal server error", http.StatusInternalServerError)
		}
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(book)
}

// updateBook handles PUT /books/{id}
func (h *Handler) updateBook(w http.ResponseWriter, r *http.Request, id int64) {
	var req UpdateBookRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		h.writeError(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	// Validate that at least one field is being updated
	if req.Title == nil && req.Author == nil && req.Year == nil && req.ISBN == nil {
		h.writeError(w, "At least one field must be provided for update", http.StatusBadRequest)
		return
	}

	// Validate title if provided
	if req.Title != nil {
		if *req.Title == "" {
			h.writeError(w, "Title cannot be empty", http.StatusBadRequest)
			return
		}
		if !isPrintableASCII(*req.Title) {
			h.writeError(w, "Title must contain only printable characters", http.StatusBadRequest)
			return
		}
		if utf8.RuneCountInString(*req.Title) > 500 {
			h.writeError(w, "Title must not exceed 500 characters", http.StatusBadRequest)
			return
		}
	}

	// Validate author if provided
	if req.Author != nil {
		if *req.Author == "" {
			h.writeError(w, "Author cannot be empty", http.StatusBadRequest)
			return
		}
		if !isPrintableASCII(*req.Author) {
			h.writeError(w, "Author must contain only printable characters", http.StatusBadRequest)
			return
		}
		if utf8.RuneCountInString(*req.Author) > 200 {
			h.writeError(w, "Author must not exceed 200 characters", http.StatusBadRequest)
			return
		}
	}

	book, err := h.store.UpdateBook(id, req)
	if err != nil {
		if strings.Contains(err.Error(), "not found") {
			h.writeError(w, err.Error(), http.StatusNotFound)
		} else {
			log.Printf("Error updating book: %v", err)
			h.writeError(w, "Internal server error", http.StatusInternalServerError)
		}
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(book)
}

// deleteBook handles DELETE /books/{id}
func (h *Handler) deleteBook(w http.ResponseWriter, r *http.Request, id int64) {
	err := h.store.DeleteBook(id)
	if err != nil {
		if strings.Contains(err.Error(), "not found") {
			h.writeError(w, err.Error(), http.StatusNotFound)
		} else {
			log.Printf("Error deleting book: %v", err)
			h.writeError(w, "Internal server error", http.StatusInternalServerError)
		}
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

// validateCreateBookRequest validates a create book request
func validateCreateBookRequest(req CreateBookRequest) error {
	if req.Title == "" {
		return fmt.Errorf("Title is required")
	}
	if req.Author == "" {
		return fmt.Errorf("Author is required")
	}
	if !isPrintableASCII(req.Title) {
		return fmt.Errorf("Title must contain only printable characters")
	}
	if !isPrintableASCII(req.Author) {
		return fmt.Errorf("Author must contain only printable characters")
	}
	if utf8.RuneCountInString(req.Title) > 500 {
		return fmt.Errorf("Title must not exceed 500 characters")
	}
	if utf8.RuneCountInString(req.Author) > 200 {
		return fmt.Errorf("Author must not exceed 200 characters")
	}
	return nil
}

// isPrintableASCII checks if a string contains only printable ASCII characters
func isPrintableASCII(s string) bool {
	for _, r := range s {
		if r < 32 || r > 126 {
			return false
		}
	}
	return true
}

// writeError writes a JSON error response
func (h *Handler) writeError(w http.ResponseWriter, message string, status int) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(ErrorResponse{Error: message})
}

File: experiments-local/experiment-15/runs/agent=omp_language=rust_model=openrouter/tencent/hy3-preview_tooling=none/rep2/src/handlers.rs
use crate::db::Database;
use crate::models::{
    BookResponse, CreateBookRequest, ErrorResponse, HealthResponse, UpdateBookRequest,
};
use actix_web::{web, HttpResponse, Responder};
use chrono::Utc;
use validator::Validate;

pub async fn health() -> impl Responder {
    HttpResponse::Ok().json(HealthResponse {
        status: "healthy".to_string(),
        timestamp: Utc::now(),
    })
}

pub async fn create_book(
    db: web::Data<Database>,
    req: web::Json<CreateBookRequest>,
) -> impl Responder {
    if let Err(e) = req.validate() {
        return HttpResponse::BadRequest().json(ErrorResponse {
            error: "ValidationError".to_string(),
            message: format!("{}", e),
        });
    }

    match db
        .create_book(&req.title, &req.author, req.year, req.isbn.as_deref())
        .await
    {
        Ok(book) => HttpResponse::Created().json(BookResponse::from(book)),
        Err(e) => HttpResponse::InternalServerError().json(ErrorResponse {
            error: "DatabaseError".to_string(),
            message: format!("{}", e),
        }),
    }
}

pub async fn list_books(
    db: web::Data<Database>,
    query: web::Query<std::collections::HashMap<String, String>>,
) -> impl Responder {
    let author_filter = query.get("author").map(|s| s.as_str());

    match db.list_books(author_filter).await {
        Ok(books) => {
            let responses: Vec<BookResponse> = books.into_iter().map(BookResponse::from).collect();
            HttpResponse::Ok().json(responses)
        }
        Err(e) => HttpResponse::InternalServerError().json(ErrorResponse {
            error: "DatabaseError".to_string(),
            message: format!("{}", e),
        }),
    }
}

pub async fn get_book(
    db: web::Data<Database>,
    path: web::Path<String>,
) -> impl Responder {
    let id = path.into_inner();

    match db.get_book(&id).await {
        Ok(book) => HttpResponse::Ok().json(BookResponse::from(book)),
        Err(sqlx::Error::RowNotFound) => HttpResponse::NotFound().json(ErrorResponse {
            error: "NotFound".to_string(),
            message: format!("Book with id {} not found", id),
        }),
        Err(e) => HttpResponse::InternalServerError().json(ErrorResponse {
            error: "DatabaseError".to_string(),
            message: format!("{}", e),
        }),
    }
}

pub async fn update_book(
    db: web::Data<Database>,
    path: web::Path<String>,
    req: web::Json<UpdateBookRequest>,
) -> impl Responder {
    let id = path.into_inner();

    if let Err(e) = req.validate() {
        return HttpResponse::BadRequest().json(ErrorResponse {
            error: "ValidationError".to_string(),
            message: format!("{}", e),
        });
    }

    // Check if book exists
    if let Err(sqlx::Error::RowNotFound) = db.get_book(&id).await {
        return HttpResponse::NotFound().json(ErrorResponse {
            error: "NotFound".to_string(),
            message: format!("Book with id {} not found", id),
        });
    }

    let isbn_update = req.isbn.as_deref();

    match db
        .update_book(
            &id,
            req.title.as_deref(),
            req.author.as_deref(),
            req.year,
            isbn_update,
        )
        .await
    {
        Ok(book) => HttpResponse::Ok().json(BookResponse::from(book)),
        Err(e) => HttpResponse::InternalServerError().json(ErrorResponse {
            error: "DatabaseError".to_string(),
            message: format!("{}", e),
        }),
    }
}

pub async fn delete_book(
    db: web::Data<Database>,
    path: web::Path<String>,
) -> impl Responder {
    let id = path.into_inner();

    // Check if book exists
    if let Err(sqlx::Error::RowNotFound) = db.get_book(&id).await {
        return HttpResponse::NotFound().json(ErrorResponse {
            error: "NotFound".to_string(),
            message: format!("Book with id {} not found", id),
        });
    }

    match db.delete_book(&id).await {
        Ok(_) => HttpResponse::NoContent().finish(),
        Err(e) => HttpResponse::InternalServerError().json(ErrorResponse {
            error: "DatabaseError".to_string(),
            message: format!("{}", e),
        }),
    }
}
